AOTAutograd: add config to error when overlapping input checks would cause slow compile / runtimes

bdhirsh opened the pull request:
We should eventually make the non-overlapping checks faster when dynamic shapes are enabled, but this is pretty difficult to do. So for now this PR adds a config that lets us fail fast when this situation happens, instead of causing compile times to secretly come to a crawl.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #123455



cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang

bdhirsh committed:
AOTAutograd: add config to error when overlapping input checks would cause slow compile / runtimes

[ghstack-poisoned]

bdhirsh committed:
Update on "AOTAutograd: add config to error when overlapping input checks would cause slow compile / runtimes"

We should eventually make the non-overlapping checks faster when dynamic shapes are enabled, but this is pretty difficult to do. So for now this PR adds a config that lets us fail fast when this situation happens, instead of causing compile times to secretly come to a crawl.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

ezyang reviewed (approved):
5 might be too low, but I guess it will be really obvious if it is

ezyang commented:
Consider JK guarding the default here

bdhirsh committed:
Update on "AOTAutograd: add config to error when overlapping input checks would cause slow compile / runtimes"

We should eventually make the non-overlapping checks faster when dynamic shapes are enabled, but this is pretty difficult to do. So for now this PR adds a config that lets us fail fast when this situation happens, instead of causing compile times to secretly come to a crawl.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

bdhirsh commented on torch/_functorch/config.py:
```diff
@@ -26,6 +27,26 @@
 
 debug_partitioner = os.environ.get("AOT_PARTITIONER_DEBUG", False)
 
+# Today, if you are in a situation where there is "false aliasing"
```
@ezyang my attempt at using JK is here, with the corresponding knob [here](https://www.internalfb.com/intern/justknobs/?name=pytorch%2Fdynamo) (I threw it under dynamo configs because a dedicated namespace felt like overkill.

I was also paranoid about the fact that `getval_int()` defaults to zero if JK isn't available for some reason. LMK if this looks reasonable to you (I cargo-culted this a bit from inductor's JK usage)

ezyang replied:
Meh, don't do it this way. Do a boolean JK and just have it be a killswitch. (When JK is not available, the killswitch should default to "not killed", because the steadystate is for the killswitch to be off.)

bdhirsh committed:
Update on "AOTAutograd: add config to error when overlapping input checks would cause slow compile / runtimes"

We should eventually make the non-overlapping checks faster when dynamic shapes are enabled, but this is pretty difficult to do. So for now this PR adds a config that lets us fail fast when this situation happens, instead of causing compile times to secretly come to a crawl.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

bdhirsh committed:
Update on "AOTAutograd: add config to error when overlapping input checks would cause slow compile / runtimes"

We should eventually make the non-overlapping checks faster when dynamic shapes are enabled, but this is pretty difficult to do. So for now this PR adds a config that lets us fail fast when this situation happens, instead of causing compile times to secretly come to a crawl.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

bdhirsh committed:
Update on "AOTAutograd: add config to error when overlapping input checks would cause slow compile / runtimes"

We should eventually make the non-overlapping checks faster when dynamic shapes are enabled, but this is pretty difficult to do. So for now this PR adds a config that lets us fail fast when this situation happens, instead of causing compile times to secretly come to a crawl.




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh commented:
@pytorchbot merge -f "trunk failure: https://ossci-raw-job-status.s3.amazonaws.com/log/23749911638"
